### ISSUE TITLE
[UX][D9] Unfiltered Text area (multiple rows) doesn't allow line breaks in spite of being labeled "Plain text" 

### ISSUE THREAD

**Issue by kswan** [backdrop/backdrop-issues#5515]
## Description of the bug

When using a multiline (text area) field with plain text processing, it is expected that line breaks will apply to the rendered page, but they don't. Drupal 8/9 inserts `<br />` tags.

## Steps To Reproduce

1. Add a field with a "Text (long)" field type and "Text area (multiple rows)" widget.
2. Set the Text processing to "Plain text"
3. Create content using the new field with multiple lines of text.
![image](https://user-images.githubusercontent.com/154391/154611150-72b97693-6a8a-41a2-b63b-e4e27dbc9b15.png)

## Actual behavior

The field is rendered without line breaks.
![image](https://user-images.githubusercontent.com/154391/154611250-35464fc7-cb82-495f-8d7b-8c6f97ae77c5.png)

## Expected behavior

The line breaks should be applied to the rendered page.
![image](https://user-images.githubusercontent.com/154391/154611696-13a495b4-2cee-4c8e-9a5e-7c217f8a35bf.png)

## Additional information

Add any other information that could help, such as:
- Backdrop CMS version: 1.21.2

**Comment by kswan** [comment]
I should clarify that this issue is related to Plain text _processing_ not the Plain text _format_.

**Comment by indigoxela** [comment]
@kswan many thanks for taking the time to file a pull request.

However, I don't agree with your approach. I adds processing that belongs to filters (formats) to a function that's only supposed to do sanitizing. With nl2br() you turn it into html, but it's supposed to be plain text.
If you need a format, you're supposed to add a filter that does what you need.

**Comment by indigoxela** [comment]
Here's some explanation of what I think you misunderstood.

There's a "plain text" processing format, but you can not choose that in the select list. (Questionable, but that's how it is in B...)

![only-fallback](https://user-images.githubusercontent.com/842066/154797516-8341dea7-efa2-48ec-b559-8668426eecf4.png)

If you want a format with line breaks (as html), you have to create a custom one which does the nl2br (and removes all other markup before that). An example for such a "No markup" format:

![custom-plain-text](https://user-images.githubusercontent.com/842066/154797550-32d49e55-7228-4b46-8930-13d1a41dbab0.png)

**Comment by BWPanda** [comment]
@kswan I suggested opening an issue for this when you said Drupal did things differently, but I tested this on one of my D7 sites, and it appears to work the same as Backdrop...

![image](https://user-images.githubusercontent.com/2385329/154798854-b8b59e30-50ae-4fff-8c66-874dc8874b71.png)

![image](https://user-images.githubusercontent.com/2385329/154798862-8cec2e86-d46c-4981-96f2-fb6b664027e9.png)

So I'm with @indigoxela that this is by design.

**Comment by kswan** [comment]
@indigoxela, I agree that the Plain text processing setting is actually a format with less markup than the Plain text format. 

My main argument for this change is that a multiline field with plain text processing is counter intuitive to most users. Also, I want to control a field to multiline plain text content (only br tags), but currently the only options are no tags (unusable as a multiline field) or text format based on the users role. The proposal of a "No markup" format would work except I can't specify that format for a field.

I guess what I really want is an option to specify the allowed text formats per field.

@BWPanda, when I tested Drupal it was version 8 or 9. This issue (https://www.drupal.org/project/drupal/issues/1152216) was resolved for Drupal 8 and there was discussion about backporting to D7, but it didn't happen.

@indigoxela and @BWPanda Is there a use case for a multiline plain text field where you would expect/want the text to be rendered without linefeeds? It seems the current version causes there to be no use cases for a multiline plain text field.

**Comment by BWPanda** [comment]
> I guess what I really want is an option to specify the allowed text formats per field.

https://github.com/backdrop-contrib/preferred_text_format

**Comment by indigoxela** [comment]
> I can't specify that format for a field.

@kswan core lets you set the default, but doesn't _enforce_ it by default. There are contrib modules for that. @BWPanda already mentioned preferred_text_format, but you might also want to have a look at [simplify](https://backdropcms.org/project/simplify).

The confusing thing might be that the *format* Plain text (which isn't selectible via dropdown) seems to do the same as no processing at all (the Plain text setting), but they differ. IMO that naming isn't optimal, but that's not the main issue here.

> Is there a use case for a multiline plain text field where you would expect/want the text to be rendered without linefeeds?

Yes, use that if you want *no* markup at all - also no line breaks. This can make sense if you need it for simple messages (one-liners), or html attributes that can't contain any markup.

**Comment by BWPanda** [comment]
> but you might also want to have a look at [simplify](https://backdropcms.org/project/simplify).

Ha ha, both my modules! 😄

**Comment by kswan** [comment]
@indigoxela  and @BWPanda I appreciate your suggestions.

I still believe this issue and PR are valid and should be committed. Drupal 8 and 9 implement this behavior and Drupal 7 probably would except the Drupal community had moved past Drupal 7 by the time the behavior was addressed. Also, I don't follow @indigoxela's use case for a "one-liner" using a "Text area (**multiple rows**)" instead of a Text field.

With that said, I am able to work around this issue using simplify and setting the text format in the Default value setting. I am glad to find simplify for other uses beyond this issue (thank you, @BWPanda and @indigoxela ).

**Comment by indigoxela** [comment]
> I still believe this issue and PR are valid and should be committed.

@kswan I respect your point of view. However, the approach chosen for the PR isn't valid. It abuses a sanitize function to turn something that's supposed to be plain text (= is not allowed to contain any markup) into HTML.
If you want markup (like automatic `<br>` and `<p>`), you need text processing.

If you want to enforce a format, simplify is the proper solution. (If you want a specific format - and even if people would have access to other formats, they shouldn't have the ability to switch.)

**Comment by herbdool** [comment]
I can confirm that Drupal 9 will add line breaks. I'm not sure if we want to add a potentially backwards breaking change, but I'm open to different approaches.

It seems that in `_text_sanitize()` it could perhaps use a new function instead of `check_plain()` as they fallback to sanitize.

**Comment by purohitmansi** [comment]
Hello @kswan!

I am an openforce 2022 contributor. I would love to work on this issue and make PR as soon as I resolve the issue. Please do assign this to me.

ThankYou:)

**Comment by indigoxela** [comment]
Hi @purohitmansi,

many thanks for trying to help. However, this isn't an easy task and actually not marked as "good first issue". The direction here is still unclear, as it would break backwards compatibility.

**Comment by BWPanda** [comment]
Marking this as a feature request since functionality we don't have is being requested.

**Comment by kswan** [comment]
Thank you, @klonos for adding support to this issue. I believe there are 2 UX WTF issues here.

First, the issue that you quoted is:
> Is there a use case for a multiline plain text field where you would expect/want the text to be rendered without linefeeds?

Also, there is overlapping usage of the term "Plain text". Based on the explanation from @indigoxela, I have distinguished them as Plain text _processing_ and Plain text _format_, but that distinction is not intuitive. I think this is the UX WTF that @echozone and I ran into.

**Comment by jenlampton** [comment]
I agree with @indigoxela  and @BWPanda - Unprocessed, `Plain text` fields should not be processed to allow line-breaks. If people want to use line-breaks, they should use `Filtered` text fields, with the `Plain text` format, which is configurable.

This is not a bug, and does work as designed, but we do still have some work to do on the UX! :)

> Also, there is overlapping usage of the term "Plain text". [...] Plain text processing and Plain text format

Yes, this is a problem :(  If we renamed one of these that would clear up some confusion. I my instinct is to leave the text format as `Plain text` and change the processing value to something like `Unprocessed`, `Unfiltered`, or `Unformatted`.

>  I don't follow @indigoxela's use case for a "one-liner" using a "Text area (multiple rows)" instead of a Text field.

I do this all the time. The reason we need to use a text-area is because the "one-liner" is often more than 80 characters long. Editors need to be able to see their text wrap to the next line, in the same way it will wrap when it's displayed. 

> The direction here is still unclear, as it would break backwards compatibility.

It's possible to create a contrib module that would process the unprocessed fields (adding label). But I don't think it's something we should do in core, since we have already met this use-case in core.

**Comment by ericfoy** [comment]
> Is there a use case for a multiline plain text field where you would expect/want the text to be rendered without linefeeds?

Interestingly, no one has really answered this. I would argue that for most site builders, the expectation would be that "plaintext" means "like Notepad - Not rich text." In literally everyone's experience, lines usually wrap at the right edge of the window in which they are displayed unless you insert a hard return. Even in the devel world, a string can traditionally hold "\n" (and other escape sequences) *apart* from any reference to html—indeed: long before html was even a thing. 

So whereas "plaintext", to the modern developer, means "no html whatsoever"... the average site builder is not unreasonable in his expectation that it could nonetheless contain line breaks. The fact that we have to sneak in the `<br>` tags in order to emulate the existence of the expected "\n" in a long text string should be our moral dilemma, not his.

It seems to me that an additional "Line Breaks Only" format could be preconfigured in the OOTB Backdrop install, which would massively help people along in discovering the format they're looking for.

**Comment by indigoxela** [comment]
I belief, what confuses people here is a misunderstanding. The actual newlines `\n` _are_ there - they just don't have any meaning inside HTML, so they _seem_ to be "gone" - they aren't.
To get a _visible_ newline in HTML requires markup. Either `<br>` or wrapped in `<p>` or the whole thing wrapped in `<pre>`.
But using markup like that isn't "plain text" - to be pedantic. 😉

The solution - while possibly not obvious - is to create a format with the "Convert line breaks into HTML" filter enabled.

**Comment by ericfoy** [comment]
> The actual newlines `\n` _are_ there - they just don't have any meaning inside HTML, so they _seem_ to be "gone" - they aren't.

Aaahh, very succinctly put. So: The issue is that of presentation, not string value storage!

Here's my proposed solution: since "plain text" is universally understood to include control characters ( https://www.google.com/search?q=can+plain+text+include+control+characters%3F&oq=can+plain+text+include+control+characters%3F ) then I would recommend that the core UI provide a selection beneath the Plain text format: "Observe line breaks" with help text: "Deselect this to display the field contents entirely inline, without line breaks." The same options should be available on the Display side—as the aforementioned UI just sets up the default.

Then, upon output of the field, BD would selectively wrap the field inside `<span style="white-space: pre-wrap;">your text here</span>`.

This would make it all make sense: It's plain text; there is no markup; and it acts like plaintext should act.

Optional... Let the user select

- Display line breaks and wrap long lines. (emits style="white-space: pre-wrap;")
- Display line breaks, but do not wrap long lines (emits style="white-space: pre-line;")
- Do not display line breaks. All text will be in one line

The OOTB setting should be the first one, as this behaves most traditionally like plain text.

With this solution, there is no confusion about markup being in there or not—there is none. So no consideration of filtering needed!

An alternate method: set up system-level css classes "plain-text-wrap" and "plain-text-line", making the markup a little cleaner (for those who cringe at inline styles) PLUS giving us the opportunity to override them if desired.

If there is any traction here, I'm happy to work up a proposed patch/PR —or— a contrib module.

**Comment by ericfoy** [comment]
I have created a module that pretty much adds the functionality described above.
https://github.com/ericfoy/better_plaintext

Note that I was mistaken about the "not wrapping long lines" thing. Instead it was "compress extra white-space."
Anyway, I would love to get feedback from the community on this. I think it clears up a lot of confusion...

Cheers,
—Eric

**Comment by NormPlum** [comment]
It seems to me that the problem is the wording of the "Plain text" processor. Is there a situation where using the "Plain text" format (or a custom one) in order to keep line breaks (but add no other HTML tags) wouldn't work?

I made a PR that renames things to hopefully make more sense:

<img width="473" height="288" alt="Image" src="https://github.com/user-attachments/assets/78dcb0b8-da29-4791-9faf-c8a2af0b1fd7" />

**Comment by swampopus** [comment]
@NormPlum I love your solution.  Just calling it "unformatted text" with a description of "No HTML code, line breaks, etc." is so much cleaner and obvious what it does.

The "Plain text" vs "Plain text" (under Filtered text) has been the bane of my existence for **at least 20 years** of Drupal and now Backdrop development:

<img width="484" height="330" alt="Image" src="https://github.com/user-attachments/assets/f1717c71-a3f6-4d94-9985-8b14f182135c" />

My solution has been to go into the text formats and rename Plain text to "Plain text w/ line breaks", like so:

<img width="598" height="272" alt="Image" src="https://github.com/user-attachments/assets/090c607f-84ec-4926-b09d-66d1236d337d" />

There's lots of little things like that I've considered bundling into a single module which I would just enable with every new site.  If your PR gets rejected, I nominate creating such a module.

**Comment by olafgrabienski** [comment]
> **Text processing**
> ( ) Unformatted text
> No HTML code, line breaks, etc.

I alsol like the new label "Unformatted text" in the PR by @NormPlum. The description ("No HTML ... etc.") sounds a bit vague to me. I don't have a real alternative at the moment, but to be more clear, I'd suggest to add another "no" before "line breaks", and to remove the "etc":

> No HTML code, no line breaks.

Hoping for better ideas by others.

**Comment by swampopus** [comment]
You know-- I've had some time to think about it.

I actually use _real_ Plain Text all the time if I'm storing data which isn't really meant to be displayed at all.  I think the key word there is "displayed."  Because the text saved with Plain Text does indeed allow HTML tags and line breaks, they just won't be displayed that way.

So here's a possible suggestion:

> ( ) **Unformatted text**
> _HTML and line breaks not displayed in web browser._

I know it's a little more wordy, but it's more precise.
